### PR TITLE
fix(swaps): ensure correct takerCltvDelta and makerCltvDelta usage

### LINKED ISSUE #984: LTC/BTC swap failure - Taker's xud crashed
Failure to take an order. 
Notes:
1.  no problem with sanity swaps.
2. LTC and BTC have direct channels
2. Error on make side atter receiving the first leg.
3. XUD crashed on the taker side.

Taker issues a buy order:
```
xud@xud-test-2:/opt/xud$ xucli listorders
Trading pair: LTC/BTC
┌────────────────────────────────────────────────────┬────────────────────────────────────────────────────┐
│ Buy                                                │ Sell                                               │
├───────────────────┬───────────────────┬────────────┼───────────────────┬───────────────────┬────────────┤
│ Quantity          │ Price             │ Own Order  │ Quantity          │ Price             │ Own Order  │
├───────────────────┼───────────────────┼────────────┼───────────────────┼───────────────────┼────────────┤
│ 3.50000000        │ 0.0077            │            │ 2.50000000        │ 0.0079            │            │
├───────────────────┼───────────────────┼────────────┼───────────────────┼───────────────────┼────────────┤
│ 15.50000000       │ 0.0076            │            │ 13.00000000       │ 0.008             │            │
├───────────────────┼───────────────────┼────────────┼───────────────────┼───────────────────┼────────────┤
│ 18.00000000       │ 0.0075            │            │ 15.60000000       │ 0.0081            │            │
├───────────────────┼───────────────────┼────────────┼───────────────────┼───────────────────┼────────────┤
│ 21.25000000       │ 0.0074            │            │ 18.10000000       │ 0.0082            │            │
├───────────────────┼───────────────────┼────────────┼───────────────────┼───────────────────┼────────────┤
│ 24.00000000       │ 0.0073            │            │ 22.30000000       │ 0.0083            │            │
└───────────────────┴───────────────────┴────────────┴───────────────────┴───────────────────┴────────────┘
Trading pair: WETH/BTC
┌────────────────────────────────────────────────────┬────────────────────────────────────────────────────┐
│ Buy                                                │ Sell                                               │
├───────────────────┬───────────────────┬────────────┼───────────────────┬───────────────────┬────────────┤
│ Quantity          │ Price             │ Own Order  │ Quantity          │ Price             │ Own Order  │
└───────────────────┴───────────────────┴────────────┴───────────────────┴───────────────────┴────────────┘
xud@xud-test-2:/opt/xud$ xucli buy 0.1 LTC/BTC 0.0079 --stream
```

Response:
```
xud@xud-test-2:/opt/xud$ xucli buy 0.1 LTC/BTC 0.0079 --stream
failed to swap 0.1 LTC with peer order d96cfab0-7f2b-11e9-9040-7b281a4ac93a
events.js:174
      throw er; // Unhandled 'error' event
      ^
Error: 14 UNAVAILABLE: TCP Read failed
    at Object.exports.createStatusError (/opt/xud/node_modules/grpc/src/common.js:91:15)
    at ClientReadableStream._emitStatusIfDone (/opt/xud/node_modules/grpc/src/client.js:233:26)
    at ClientReadableStream._receiveStatus (/opt/xud/node_modules/grpc/src/client.js:211:8)
    at Object.onReceiveStatus (/opt/xud/node_modules/grpc/src/client_interceptors.js:1272:15)
    at InterceptingListener._callNext (/opt/xud/node_modules/grpc/src/client_interceptors.js:568:42)
    at InterceptingListener.onReceiveStatus (/opt/xud/node_modules/grpc/src/client_interceptors.js:618:8)
    at /opt/xud/node_modules/grpc/src/client_interceptors.js:1029:24
Emitted 'error' event at:
    at ClientReadableStream._emitStatusIfDone (/opt/xud/node_modules/grpc/src/client.js:234:12)
    at ClientReadableStream._receiveStatus (/opt/xud/node_modules/grpc/src/client.js:211:8)
    [... lines matching original stack trace ...]
    at /opt/xud/node_modules/grpc/src/client_interceptors.js:1029:24
```

Taker log:
```
25/05/2019 20:31:13.890 [ORDERBOOK] debug: reduced order d96cfab0-7f2b-11e9-9040-7b281a4ac93a by 10000000 quantity while matching order 0a41be00-7f2c-11e9-9a17-fb1c5525c06e
25/05/2019 20:31:13.892 [ORDERBOOK] debug: matched with peer 02b66438730d1fcdf4a4ae5d3d73e847a272f160fee2938e132b52cab0a0d9cfc6, executing swap on taker 0a41be00-7f2c-11e9-9a17-fb1c552
5c06e and maker d96cfab0-7f2b-11e9-9040-7b281a4ac93a for 10000000
25/05/2019 20:31:13.922 [LND] debug: BTC: got 1 route(s) to destination 02d693d951ce4fc116b2b5e22e8e4966dd3cf9ad7600b8de24190d6af429f260c0: 291915,,79000,320796811054874600,2500000000,
79000,,291915,79000000,,02d693d951ce4fc116b2b5e22e8e4966dd3cf9ad7600b8de24190d6af429f260c0,,79000000
25/05/2019 20:31:13.924 [SWAPS] debug: New deal: {"takerCltvDelta":576,"rHash":"81edc66eb9587c37596b629ef9d96a70ff1d0d9601d6bfbdead409f071e24603","orderId":"d96cfab0-7f2b-11e9-9040-7b2
81a4ac93a","pairId":"LTC/BTC","proposedQuantity":10000000,"rPreimage":"9e8e237ddb0804b9fca1673ccfa9fc33eb4de70392a56c38a21732e0f56e2efa","takerCurrency":"LTC","makerCurrency":"BTC","ta
kerAmount":10000000,"makerAmount":79000,"destination":"02d693d951ce4fc116b2b5e22e8e4966dd3cf9ad7600b8de24190d6af429f260c0","peerPubKey":"02b66438730d1fcdf4a4ae5d3d73e847a272f160fee2938
e132b52cab0a0d9cfc6","localId":"0a41be00-7f2c-11e9-9a17-fb1c5525c06e","price":0.0079,"isBuy":false,"phase":0,"state":0,"role":0,"createTime":1558816273923}
25/05/2019 20:31:13.926 [SWAPS] debug: Requesting deal: {"takerCltvDelta":576,"rHash":"81edc66eb9587c37596b629ef9d96a70ff1d0d9601d6bfbdead409f071e24603","orderId":"d96cfab0-7f2b-11e9-9
040-7b281a4ac93a","pairId":"LTC/BTC","proposedQuantity":10000000,"rPreimage":"9e8e237ddb0804b9fca1673ccfa9fc33eb4de70392a56c38a21732e0f56e2efa","takerCurrency":"LTC","makerCurrency":"B
TC","takerAmount":10000000,"makerAmount":79000,"destination":"02d693d951ce4fc116b2b5e22e8e4966dd3cf9ad7600b8de24190d6af429f260c0","peerPubKey":"02b66438730d1fcdf4a4ae5d3d73e847a272f160
fee2938e132b52cab0a0d9cfc6","localId":"0a41be00-7f2c-11e9-9a17-fb1c5525c06e","price":0.0079,"isBuy":false,"phase":0,"state":0,"role":0,"createTime":1558816273923}
25/05/2019 20:31:13.982 [P2P] debug: received swapAccepted from 02b66438730d1fcdf4a4ae5d3d73e847a272f160fee2938e132b52cab0a0d9cfc6: {"rHash":"81edc66eb9587c37596b629ef9d96a70ff1d0d9601
d6bfbdead409f071e24603","quantity":10000000,"makerCltvDelta":288}
25/05/2019 20:31:13.992 [LND] debug: LTC: added invoice of 10000000 for 81edc66eb9587c37596b629ef9d96a70ff1d0d9601d6bfbdead409f071e24603
25/05/2019 20:31:14.447 [LND] error: BTC: sendPaymentSync failed with payment error: UnknownPaymentHash(amt=79000000 mSAT)
25/05/2019 20:31:14.450 [SWAPS] debug: deal 81edc66eb9587c37596b629ef9d96a70ff1d0d9601d6bfbdead409f071e24603 failed due to SendPaymentFailure
25/05/2019 20:31:14.457 [ORDERBOOK] error: swap between orders d96cfab0-7f2b-11e9-9040-7b281a4ac93a & 0a41be00-7f2c-11e9-9a17-fb1c5525c06e failed due to SendPaymentFailure
**25/05/2019 20:31:25.563 [GLOBAL] info: config loaded**
25/05/2019 20:31:25.701 [DB] info: connected to database /home/xud/.xud/xud-simnet.db
25/05/2019 20:31:25.727 [RAIDEN] info: RaidenClient status: Disabled
25/05/2019 20:31:25.730 [LND] info: BTC: trying to verify connection to lnd at localhost:10002
25/05/2019 20:31:25.736 [LND] info: LTC: trying to verify connection to lnd at localhost:10001
25/05/2019 20:31:25.759 [LND] info: BTC: LndClient status: ConnectionVerified
```

Maker log:
```
25/05/2019 20:31:13.932 [P2P] debug: received swapRequest from 028599d05b18c0c3f8028915a17d603416f7276c822b6b2d20e71a3502bd0f9e0a: {"proposedQuantity":10000000,"pairId":"LTC/BTC","orde
rId":"d96cfab0-7f2b-11e9-9040-7b281a4ac93a","rHash":"81edc66eb9587c37596b629ef9d96a70ff1d0d9601d6bfbdead409f071e24603","takerCltvDelta":576}
25/05/2019 20:31:13.934 [ORDERBOOK] debug: added hold of 10000000 on order d96cfab0-7f2b-11e9-9040-7b281a4ac93a
25/05/2019 20:31:13.936 [SWAPS] debug: New deal: {"proposedQuantity":10000000,"pairId":"LTC/BTC","orderId":"d96cfab0-7f2b-11e9-9040-7b281a4ac93a","rHash":"81edc66eb9587c37596b629ef9d96
a70ff1d0d9601d6bfbdead409f071e24603","takerCltvDelta":576,"takerPubKey":"0276e56e5525602dc5456ee55c2e28f201661085bce7f0064400ed83f0c167e8ef","price":0.0079,"isBuy":false,"quantity":100
00000,"makerAmount":79000,"takerAmount":10000000,"makerCurrency":"BTC","takerCurrency":"LTC","destination":"0276e56e5525602dc5456ee55c2e28f201661085bce7f0064400ed83f0c167e8ef","peerPub
Key":"028599d05b18c0c3f8028915a17d603416f7276c822b6b2d20e71a3502bd0f9e0a","localId":"d96cfab0-7f2b-11e9-9040-7b281a4ac93a","phase":0,"state":0,"role":1,"createTime":1558816273935}
25/05/2019 20:31:13.942 [LND] debug: LTC: got 1 route(s) to destination 0276e56e5525602dc5456ee55c2e28f201661085bce7f0064400ed83f0c167e8ef: 14681,,10000000,12244161486979072,2500000000
0,10000000,,14681,10000000000,,0276e56e5525602dc5456ee55c2e28f201661085bce7f0064400ed83f0c167e8ef,,10000000000
25/05/2019 20:31:13.957 [SWAPS] debug: got block height of 14105
25/05/2019 20:31:13.957 [SWAPS] debug: total timelock of route = 576 makerCltvDelta = 288
25/05/2019 20:31:13.976 [LND] debug: BTC: added invoice of 79000 for 81edc66eb9587c37596b629ef9d96a70ff1d0d9601d6bfbdead409f071e24603
25/05/2019 20:31:13.979 [SWAPS] debug: Sending swap response to peer 
25/05/2019 20:31:14.213 [SWAPS] debug: Executing maker code to resolve hash
25/05/2019 20:31:14.262 [LND] error: LTC: sendToRouteSync failed with payment error: unable to route payment to destination: UnknownNextPeer
25/05/2019 20:31:14.264 [SWAPS] debug: deal 81edc66eb9587c37596b629ef9d96a70ff1d0d9601d6bfbdead409f071e24603 failed due to SendPaymentFailure
25/05/2019 20:31:14.268 [ORDERBOOK] debug: removed hold of 10000000 on order d96cfab0-7f2b-11e9-9040-7b281a4ac93a
25/05/2019 20:31:14.263 [LND] error: LTC: got exception from sendToRouteSync: {"paymentHash":"","paymentHashString":"81edc66eb9587c37596b629ef9d96a70ff1d0d9601d6bfbdead409f071e24603","
routesList":[{"totalTimeLock":14681,"totalFees":0,"totalAmt":10000000,"hopsList":[{"chanId":12244161486979072,"chanCapacity":25000000000,"amtToForward":10000000,"fee":0,"expiry":14681,
"amtToForwardMsat":10000000000,"feeMsat":0,"pubKey":"0276e56e5525602dc5456ee55c2e28f201661085bce7f0064400ed83f0c167e8ef"}],"totalFeesMsat":0,"totalAmtMsat":10000000000}]}
25/05/2019 20:31:14.269 [SWAPS] error: could not settle invoice: Error: Got exception from sendPaymentSync unable to route payment to destination: UnknownNextPeer
    at Swaps.<anonymous> (/opt/xud/dist/swaps/Swaps.js:504:27)
    at Generator.throw (<anonymous>)
    at rejected (/opt/xud/dist/swaps/Swaps.js:5:65)
25/05/2019 20:31:14.291 [LND] debug: BTC: canceled invoice for 81edc66eb9587c37596b629ef9d96a70ff1d0d9601d6bfbdead409f071e24603
25/05/2019 20:31:14.292 [LND] debug: BTC: deleted invoice subscription for 81edc66eb9587c37596b629ef9d96a70ff1d0d9601d6bfbdead409f071e24603
25/05/2019 20:31:14.487 [P2P] info: Peer 028599d05b18c0c3f8028915a17d603416f7276c822b6b2d20e71a3502bd0f9e0a socket closed
25/05/2019 20:31:14.489 [ORDERBOOK] debug: removed all orders for peer 028599d05b18c0c3f8028915a17d603416f7276c822b6b2d20e71a3502bd0f9e0a
25/05/2019 20:31:25.832 [P2P] debug: Connected pre-handshake to peer 35.231.171.148:47512
```

### LINKED ISSUE #1028: Digg into `SendToRoute` issue

### PR DESCRIPTION
**Pull request by ghost**
This PR fixes several bugs around `Swaps.acceptDeal` - maker's opportunity to either accept or reject the `SwapRequest`, more specifically the logic around setting the CLTV values.

- include `takerCltvDelta` as `finalCltvDelta` when querying for `makerToTakerRoutes`. This is necessary to ensure we calculate `makerCltvDelta` correctly.
- `LndClient.addInvoice` now requires `cltvExpiry` so that we can force the minimum duration of the route by specifying the `makerCltvDelta`. Previously, it was possible for the first leg to be shorter than the second one  as described in https://github.com/ExchangeUnion/xud/pull/1025#issuecomment-502047537
- remove unnecessary logging added by #1035 and #1033
- send error to peer when accepting the deal failed because something went wrong while creating the invoice. Previously we only rejected the deal and logged an error.

Closes #1028 
Closes #984

### CODE REVIEW

File: lib/lndclient/LndClient.ts [hunk 1]
```diff
@@ -352,15 +352,16 @@ class LndClient extends SwapClient {
     return this.unaryCall<lndrpc.ListChannelsRequest, lndrpc.ListChannelsResponse>('listChannels', new lndrpc.ListChannelsRequest());
   }
 
-  public getRoutes =  async (amount: number, destination: string): Promise<lndrpc.Route[]> => {
+  public getRoutes =  async (amount: number, destination: string, cltvDelta?: number): Promise<lndrpc.Route[]> => {
     const request = new lndrpc.QueryRoutesRequest();
+    const finalCltvDelta = cltvDelta || this.cltvDelta;
```
Comment (offerm) [inline lib/lndclient/LndClient.ts]: Why this.cltvDelta should be used here?
argument should not be optional IMHO

File: lib/swaps/Swaps.ts [hunk 2]
```diff
@@ -140,7 +140,7 @@ class Swaps extends EventEmitter {
       this.sanitySwaps.set(rHash, sanitySwap);
       const swapClient = this.swapClientManager.get(currency)!;
       try {
-        await swapClient.addInvoice(rHash, 1);
+        await swapClient.addInvoice(rHash, 1, swapClient.cltvDelta);
```
Comment (offerm) [inline lib/swaps/Swaps.ts]: This should be handled in the same way. The first leg should include the second leg.

File: lib/lndclient/LndClient.ts (hunk 1 above)
Comment (erkarl) [reply lib/lndclient/LndClient.ts]: I originally thought about making the argument mandatory, but the second usage of `getRoutes` happens within the context of maker at the beginning of the deal execution process when [taker queries routes to maker](https://github.com/ExchangeUnion/xud/blob/master/lib/swaps/Swaps.ts#L254).

At this point taker has the following information:
```
const makerOrder = {
  id: 'fb1708f0-90d7-11e9-8836-9bf81440c6fa',
  pairId: 'LTC/BTC',
  price: 0.1,
  quantity: 10000,
  isBuy: true,
  peerPubKey:
   '038d48a01d25735cd7bd26ac8cbd7583e4c202718848ab63b08b378dc840a24e81',
  createdAt: 1560759291688,
  initialQuantity: 10000,
};

const takerOrder = {
  pairId: 'LTC/BTC',
  price: 0.1,
  quantity: 10000,
  isBuy: false,
  localId: 'fd6da320-90d7-11e9-9e9d-0f0eb626ffe9',
  hold: 0,
  id: 'fd6da320-90d7-11e9-9e9d-0f0eb626ffe9',
  initialQuantity: 10000,
  createdAt: 1560759295570,
};
```

What would you set the `finalCltvDelta` to? Furthermore, the value of `takerToMakerRoutes` is being used as a light sanity check (error when 0 routes found). Alternative option is to remove the `getRoutes` check from the taker side.

File: lib/swaps/Swaps.ts (hunk 2 above)
Comment (erkarl) [reply lib/swaps/Swaps.ts]: In the case of `SanitySwap` and `SanitySwapAck` we're currently not exchanging CLTV delta information.

If it's OK with you I'd like to handle everything related to the sanity swaps in a separate issue/PR.

File: lib/lndclient/LndClient.ts (hunk 1 above)
Comment (erkarl) [reply lib/lndclient/LndClient.ts]: In order to be more explicit I changed `getRoutes` to require `finalCltvDelta`.

**Comment by kilrau** [comment]
Does this fix https://github.com/ExchangeUnion/xud/issues/977 ?

**Review by offerm — APPROVED** [review]

**Comment by erkarl** [comment]
I doubt it will fix #977 because this PR didn't touch sanity swaps' flow/logic. In the case of SanitySwap and SanitySwapAck we're currently not exchanging CLTV delta information and due to that it can still fail with `FinalExpiryTooSoon`.

File: lib/swaps/Swaps.ts (hunk 2 above)
Comment (sangaman) [reply lib/swaps/Swaps.ts]: I also think it's better to handle sanity swaps in a separate PR.

**Review by sangaman — APPROVED** [review]
Looks good to me, especially the new test cases. Thanks!

File: lib/lndclient/LndClient.ts (hunk 1 above)
Comment (sangaman) [reply lib/lndclient/LndClient.ts]: Fwiw I found it a bit odd to pass an object's own property to one of its method calls as in `routes = await swapClient.getRoutes(makerAmount, destination, swapClient.cltvDelta);`.

Another way to consider writing this is:

```
public getRoutes =  async (amount: number, destination: string, finalCltvDelta = this.cltvDelta) => {
```

That way it's clear what the value defaults to if its not specified. Ultimately not a big deal though so it definitely be left as is.

Comment (erkarl) [reply lib/lndclient/LndClient.ts]: After thinking about it for a while I agree, it's weird to pass an object's own property to one of its method calls. I changed it back.

**Pull request merged**